librad/src/core.rs

```diff
@@ -0,0 +1,276 @@
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use std::{
+    collections::HashMap,
+    fmt::{self, Display},
+    io,
+    path::{Path, PathBuf},
+};
+
+use async_trait::async_trait;
+use futures::Stream;
+use multibase::Base;
+use multihash::Multihash;
+use url::Url;
+
+use crate::{keys::device::Signature, peer::PeerId};
+
+/// A `RadUrn` identifies a branch in a verifiable `radicle-link` repository,
+/// where:
+///
+/// * The repository is named `id`
+/// * The initial (parent-less) revision of some identity document has the
+///   content address `id` -- where the identity document is located in the
+///   tree, and what its content schema is, is define by the [`Verifier`].
+/// * There exists a branch named `rad/id` pointing to the most recent revision
+///   of the identity document
+/// * There MAY exist a branch named `path`
+///
+/// The textual representation of a `RadUrn` is of the form:
+///
+/// ```text
+/// 'rad' ':' MULTIBASE(<id>) '/' <path>
+/// ```
+///
+/// where the preferred base is `z-base32`.
+///
+/// For example: `rad:deadbeefdeaddeafbeef/rad/issues`
+pub struct RadUrn {
+    pub id: Multihash,
+    pub path: PathBuf,
+}
+
+impl RadUrn {
+    pub fn into_rad_url(self, peer: PeerId) -> RadUrl {
+        RadUrl {
+            authority: peer,
+            urn: self,
+        }
+    }
+}
+
+impl Display for RadUrn {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "rad:{}/{}",
+            multibase::encode(Base::Base32Z, &self.id),
+            self.path.to_str().unwrap()
+        )
+    }
+}
+
+/// A `RadUrl` is a URL with the scheme `rad://`.
+///
+/// The authority of a rad URL is a [`PeerId`], from which to retrieve the
+/// `radicle-link` repository and branch identified by [`RadUrn`].
+pub struct RadUrl {
+    authority: PeerId,
+    urn: RadUrn,
+}
+
+impl RadUrl {
+    pub fn into_url(self) -> Url {
+        self.into()
+    }
+
+    // TODO: like `Gossip::fetch`, we should also be able to open a `RadUrl` from
+    // local storage:
+    // pub fn open(&self) -> Result<impl Iterator<Item = Commit>, ??>
+}
+
+impl Into<Url> for RadUrl {
+    fn into(self) -> Url {
+        let mut url = Url::parse(&format!("rad://{}", self.authority.default_encoding())).unwrap();
+
+        url.set_path(
+            Path::new(&multibase::encode(Base::Base32Z, self.urn.id))
+                .join(self.urn.path)
+                .to_str()
+                .unwrap(),
+        );
+
+        url
+    }
+}
+
+/// Placeholder for a version in a history
+type Version<'a> = &'a [u8];
```

Comment:
Because I don't know yet what the type should be :)

I _think_ that the verifier actually only needs the `revision` field, but the return type must be the VCS-native revision (ie. `Oid` for git), because we want to reset the branch to the most recent one which passes verification. Which leads us to the old problem: which backend is used is either decided at runtime, or we need to bake that assumption into the URN, so the caller can choose the type.

Reply:
Ah, I confused this with the `rad-version` in the metadata